### PR TITLE
[GEOS-11561]: Support client-delegating MapML-Proxy

### PR DESCRIPTION
**Pull request by dromagnoli**
[GEOS-11561](https://osgeo-org.atlassian.net/browse/GEOS-11561) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLBaseProxyTest.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        }
+    }
+
+    /**
+     * Executes a request using the GET method and returns the result as an MapML document.
+     *
+     * @param path The portion of the request after the context, example:
+     * @return A result of the request parsed into a dom.
+     */
+    protected org.w3c.dom.Document getMapML(final String path) throws Exception {
```
Comment (aaime) [inline src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLBaseProxyTest.java:125]: Unnecessary class name qualification?

```suggestion
    protected Document getMapML(final String path) throws Exception {
```

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLBaseProxyTest.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        MockHttpServletRequest request = createRequest(path, false);
+        request.addHeader("Accept", "text/mapml");
+        request.setMethod("GET");
+        request.setContent(new byte[] {});
+        String resp = dispatch(request, "UTF-8").getContentAsString();
+        return dom(new ByteArrayInputStream(resp.getBytes()), true);
+    }
+
+    /** For debugging purposes */
+    public static void printDocument(Document doc) throws TransformerException {
```
Comment (aaime) [inline src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLBaseProxyTest.java:135]: Why not just use the existing "GeoServerSystemTestSupport.print(Document)"  method?
It's already set up to write on System.out when running tests from IDE, but skip prints when running them over a maven build.

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSDataAccessProxyTest.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    @Override
+    protected void onSetUp(SystemTestData testData) throws Exception {
+        super.onSetUp(testData);
+        Catalog catalog = getCatalog();
+
+        WMSStoreInfo wmsStore = catalog.getFactory().createWebMapServer();
+        wmsStore.setName("wmsStore");
+        wmsStore.setWorkspace(catalog.getDefaultWorkspace());
+        wmsStore.setCapabilitiesURL(
+                "http://localhost:" + mockService.port() + getCapabilitiesURL());
```
Comment (aaime) [inline src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSDataAccessProxyTest.java:58]: Every single usage of getCapabilitiesURL is a copy/paste of this concatenation. I'd recommend merging everything inside getCapabilitiesURL then.

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSDataAccessProxyTest.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        tam.putLimits(
+                username, layer2, new DataAccessLimits(CatalogMode.CHALLENGE, intersectsFilter));
+    }
+
+    @Test
+    public void testMapMLNotCascadingWithAccessLimits() throws Exception {
+        // get the mapml doc for the layer
+        String path = BASE_REQUEST.replace("cascadedLayer", "cascadedLayerAccessLimits");
+
+        checkCascading(path, false, "image");
```
Comment (aaime) [inline src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSDataAccessProxyTest.java:106]: "image" and "tile" should probably be constants in the base class

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSDataAccessProxyTest.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        catalog.add(wmsStore);
+
+        // Create A layer with access limits
+        WMSLayerInfo wmsLayer2 = catalog.getFactory().createWMSLayer();
+        wmsLayer2.setName("cascadedLayerAccessLimits");
+        wmsLayer2.setNativeName("topp:states");
+        wmsLayer2.setStore(wmsStore);
+        wmsLayer2.setAdvertised(true);
+        wmsLayer2.setEnabled(true);
+        wmsLayer2.getMetadata().put("mapml.useRemote", true);
```
Comment (aaime) [inline src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMSDataAccessProxyTest.java:69]: There is already a constant for this one, I would use it in all tests

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMTSProxyTest.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                    + "&format_options="
+                    + MapMLConstants.MAPML_WMS_MIME_TYPE_OPTION
+                    + ":image/png";
+
+    @Override
+    protected void registerNamespaces(Map<String, String> namespaces) {
+        super.registerNamespaces(namespaces);
+        namespaces.put("wmts", "http://www.opengis.net/wmts/1.0");
+        namespaces.put("gml", "http://www.opengis.net/gml");
+        namespaces.put("xlink=", "http://www.w3.org/1999/xlink");
```
Comment (aaime) [inline src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMTSProxyTest.java:49]: Typo?

```suggestion
        namespaces.put("xlink", "http://www.w3.org/1999/xlink");
```

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMTSProxyTest.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+                    + "&SRS=MapML:OSMTILE"
+                    + "&BBOX=-1.3885038382960921E7,2870337.130793682,-7455049.489182421,6338174.0557576185"
+                    + "&WIDTH=768"
+                    + "&HEIGHT=414"
+                    + "&format_options="
+                    + MapMLConstants.MAPML_WMS_MIME_TYPE_OPTION
+                    + ":image/png";
+
+    @Override
+    protected void registerNamespaces(Map<String, String> namespaces) {
```
Comment (aaime) [inline src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMTSProxyTest.java:45]: Wondering if all the namespaces can just be declared in the base class... or are there conflicts?

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMTSProxyTest.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    public void testRemoteVsNotRemote() throws Exception {
+        Catalog cat = getCatalog();
+        // Verify the layer was added
+        LayerInfo li = cat.getLayerByName("cascadedLayer");
+        assertNotNull(li);
+        assertEquals("cascadedLayer", li.getName());
+
+        ResourceInfo layerMeta = li.getResource();
+        layerMeta.getMetadata().put("mapml.useRemote", false);
+        layerMeta.getMetadata().put("mapml.useTiles", true);
```
Comment (aaime) [inline src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLWMTSProxyTest.java:116]: This one should have a constant already available too I believe (not verified though).

**Review by aaime — CHANGES_REQUESTED** [review]
Overall looks good, see comments below

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLRequestMangler.java [hunk 9]
```diff
@@ -0,0 +1,675 @@
...
+                        "EPSG:3857"));
+        crsMappers.add(
+                new CRSMapper(
+                        Set.of("EPSG:5936", "urn:ogc:def:crs:EPSG::5936", "MapML:APSTILE"),
+                        "EPSG:5936"));
+        crsMappers.add(
+                new CRSMapper(
+                        Set.of("EPSG:3978", "urn:ogc:def:crs:EPSG::3978", "MapML:CBMTILE"),
+                        "EPSG:3978"));
+    }
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLRequestMangler.java:127]: Yep Indeed. This has been setup at the very beginning of the development, even before I was aware of the task of custom TCRS support. I will check that indeed.

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLBaseProxyTest.java (hunk 1 above)
Comment (dromagnoli) [reply src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLBaseProxyTest.java:125]: Right. No idea where did that come from :)

File: src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLBaseProxyTest.java (hunk 2 above)
Comment (dromagnoli) [reply src/extension/mapml/src/test/java/org/geoserver/mapml/MapMLBaseProxyTest.java:135]: Ah. I didn't know about that. Nice. Updating it accordingly.

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLRequestMangler.java [hunk 10]
```diff
@@ -0,0 +1,675 @@
...
+                            }
+                        }
+                    } catch (IOException e) {
+                        reason =
+                                "Unable to extract the WMS remote capabilities. Cascading won't be performed";
+                        LOGGER.warning(reason + "due to:" + e);
+                        cascadeToRemote = false;
+                    }
+                } else if (storeInfo instanceof WMTSStoreInfo) {
+                    WMTSStoreInfo wmtsStoreInfo = (WMTSStoreInfo) storeInfo;
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLRequestMangler.java:284]: There was a stashed commit, applying the same logic. I have restored it.

**Comment by prushforth** [comment]
Hi Daniele,  Sorry for this late comment.  I noticed that when I set up this wms https://agriculture.canada.ca/atlas/services/imageservices/annual_crop_inventory_2023/ImageServer/WMSServer?request=GetCapabilities&service=WMS as a proxied layer, that the GetFeatureInfo requests go to localhost not to the original domain.  Also, the localhost getfeatureinfo raises an exception because the request is missing the QUERY_LAYERS parameter:  Either no layer was queryable, or no layers were specified using QUERY_LAYERS

**Comment by aaime** [comment]
@prushforth that server does not even declare GetFeatureInfo support, in the capabilities, which also implies we don't know which formats are supported. Proxying is thus impossible. Having a fallback to the local server might not be the best option (in this case it will fail), but I also don't want to lose the option of having a working GetFeatureInfo when the remote server can do GML, but not MapML: the local GeoServer can turn that GML into whatever format available, but the MAPML client cannot use it directly, no?

**Comment by prushforth** [comment]
I thought it should not generate a map-link rel=query when the remote layer is not queryable.

**Comment by aaime** [comment]
I've discussed it with Daniele, the situation is as follows:

image

1) If the MapML client can talk directly to the remote server (queriable and supported format) it will
2) If not, but GeoServer can do a GetFeatureInfo translating the eventual remote GML format support, the client will ask GFI to the local GeoServer
3) Otherwise, no link is going to be implemented

**Comment by prushforth** [comment]
Sounds good, thanks

File: src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLDocumentBuilder.java [hunk 11]
```diff
@@ -1695,17 +1661,33 @@ private void generateWMTSQueryClientLinks(MapMLLayerMetadata mapMLLayerMetadata)
...
-                                    baseUrlPattern, path, params, URLMangler.URLType.SERVICE),
-                            "UTF-8");
-        } catch (UnsupportedEncodingException uee) {
+        MapMLRequestMangler mangler =
+                new MapMLRequestMangler(
+                        mapContent, mapMLLayerMetadata, baseUrlPattern, path, params, proj);
+        String urlTemplate = mangler.getUrlTemplate();
+        // It may be that the mangler decided to not generate any query URL due
+        // to unsupported info formats from the remote layer. So we are not
+        // generating the query link.
```
[the remark this answers is not in the document]
Comment (dromagnoli) [reply src/extension/mapml/src/main/java/org/geoserver/mapml/MapMLDocumentBuilder.java:1673]: Fixed as result of the discussion about different support types of GetFeatureInfo

**Comment by prushforth** [comment]
Thank you

**Review by aaime — APPROVED** [review]
LGTM

**Pull request merged**